In [ ]:
import torch
print(torch.__version__) # Checking version of PyTorch

2.5.1+cu124


In [ ]:
# Checking if GPU is available or not
if torch.cuda.is_available():
  print("GPU is available!")
  print(f"Using GPU: {torch.cuda.get_device_name(0)}")
else:
  print("GPU not available. Using CPU.")

GPU is available!
Using GPU: Tesla T4


## Creating Tensors

In [ ]:
# Using empty
torch.empty(2,3)

tensor([[0., 0., 0.],
        [0., 0., 0.]])

Empty function allocates memory 1st or creates spaces and shows those random values which are already there in that space.

In [ ]:
# Using empty
torch.empty(2,3)

tensor([[4.3665e-10, 4.4717e-41, 4.3665e-10],
        [4.4717e-41, 0.0000e+00, 0.0000e+00]])

In [ ]:
# Checking type
a = torch.empty(2,3)
print(type(a))

<class 'torch.Tensor'>


In [ ]:
a.dtype

torch.float32

In [ ]:
# Using zeros
torch.zeros(2,3)

tensor([[0., 0., 0.],
        [0., 0., 0.]])

In [ ]:
# Using Ones
torch.ones(2,3)

tensor([[1., 1., 1.],
        [1., 1., 1.]])

In [ ]:
# Using rand
torch.rand(2,3)

tensor([[0.9105, 0.4023, 0.9852],
        [0.2696, 0.8721, 0.4968]])

Rand will give random values from 0 to 1

In [ ]:
# Manual seed
torch.manual_seed(42)
torch.rand(2,3)

tensor([[0.8823, 0.9150, 0.3829],
        [0.9593, 0.3904, 0.6009]])

In [ ]:
torch.manual_seed(42)
torch.rand(2,3)

tensor([[0.8823, 0.9150, 0.3829],
        [0.9593, 0.3904, 0.6009]])

Manual seed helps to keep the random values constant over the time no matter how many times we run that particular code, the values will be same. Here, in this case, it will show:

tensor([[0.8823, 0.9150, 0.3829],
        [0.9593, 0.3904, 0.6009]])

In [ ]:
# Using tensor
torch.tensor([[1,2,3],
              [4,5,6],
              [7,8,9]]) # Creating custom tensors

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])

In [ ]:
# Other ways to create tensor

# arange
print("using arange ->", torch.arange(0,10,2))

# using linspace
print("using linspace ->", torch.linspace(0,10,10))

# using eye
print("using eye ->", torch.eye(5))

# using full
print("using full ->", torch.full((3, 3), 5))

using arange -> tensor([0, 2, 4, 6, 8])
using linspace -> tensor([ 0.0000,  1.1111,  2.2222,  3.3333,  4.4444,  5.5556,  6.6667,  7.7778,
         8.8889, 10.0000])
using eye -> tensor([[1., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0.],
        [0., 0., 1., 0., 0.],
        [0., 0., 0., 1., 0.],
        [0., 0., 0., 0., 1.]])
using full -> tensor([[5, 5, 5],
        [5, 5, 5],
        [5, 5, 5]])


## Tensor Shape

In [ ]:
x = torch.tensor([[1,2,3],[4,5,6]])
print(x)
print(x.shape)

tensor([[1, 2, 3],
        [4, 5, 6]])
torch.Size([2, 3])


In [ ]:
# Using empty_like
torch.empty_like(x)

tensor([[0, 0, 0],
        [0, 1, 0]])

By using this function, it will create a new tensor but with the same shape.

In [ ]:
torch.zeros_like(x)

tensor([[0, 0, 0],
        [0, 0, 0]])

In [ ]:
torch.ones_like(x)

tensor([[1, 1, 1],
        [1, 1, 1]])

In [ ]:
torch.rand_like(x,
                dtype=torch.float32)

tensor([[0.2566, 0.7936, 0.9408],
        [0.1332, 0.9346, 0.5936]])

## Tensor Datatype

In [ ]:
x.dtype

torch.int64

In [ ]:
torch.tensor([1.0,2.0,3.0],
             dtype=torch.int32)

tensor([1, 2, 3], dtype=torch.int32)

In [ ]:
torch.tensor([1,2,3],
             dtype=torch.float64)

tensor([1., 2., 3.], dtype=torch.float64)

In [ ]:
# using to()
x.to(torch.float32)

tensor([[1., 2., 3.],
        [4., 5., 6.]])

## Mathematical Operations

### Scalar operation

In [ ]:
x = torch.rand([2,2])

In [ ]:
x

tensor([[0.8694, 0.5677],
        [0.7411, 0.4294]])

In [ ]:
# Addition
print(x)
print(x+1)

tensor([[0.8694, 0.5677],
        [0.7411, 0.4294]])
tensor([[1.8694, 1.5677],
        [1.7411, 1.4294]])


### Element wise operation

In [ ]:
a = torch.rand([2,2])
b = torch.rand([2,2])

print(a)
print(b)

tensor([[0.8854, 0.5739],
        [0.2666, 0.6274]])
tensor([[0.2696, 0.4414],
        [0.2969, 0.8317]])


In [ ]:
a = torch.rand([2,2])
b = torch.rand([2,2])

print(a)
print(b)

# Addition
print(a+b)

# Subtraction
print(a-b)

tensor([[0.1053, 0.2695],
        [0.3588, 0.1994]])
tensor([[0.5472, 0.0062],
        [0.9516, 0.0753]])
tensor([[0.6525, 0.2757],
        [1.3104, 0.2746]])
tensor([[-0.4419,  0.2633],
        [-0.5927,  0.1241]])


In [ ]:
# Absolute value
c = torch.tensor([1, -8, 9, -6])

print(torch.abs(c))

tensor([1, 8, 9, 6])


In [ ]:
# Round
d = torch.tensor([1.9, 2.3, 3.7, 4.4])

torch.round(d)

tensor([2., 2., 4., 4.])

In [ ]:
# clamp
torch.clamp(d, min=2, max=3)

tensor([2.0000, 2.3000, 3.0000, 3.0000])

### Reduction Operation

In [ ]:
e = torch.randint(size=(2,3), low=0, high=10)
print(e)

# Sum
print(torch.sum(e))

# Sum along columns
print(torch.sum(e, dim=0))

# Sum along rows
print(torch.sum(e, dim=1))

tensor([[0, 9, 0],
        [9, 6, 9]])
tensor(33)
tensor([ 9, 15,  9])
tensor([ 9, 24])


## Matrix Operations

In [ ]:
# Random Matrices
f = torch.randint(size=(2,3), low=0, high=10)
g = torch.randint(size=(3,2), low=0, high=10)

print(f)
print(g)

# Matrix Multiplcation
print("Multiplication:", torch.matmul(f,g))

tensor([[1, 4, 0],
        [3, 7, 5]])
tensor([[7, 1],
        [5, 7],
        [5, 8]])
Multiplication: tensor([[27, 29],
        [81, 92]])


## Special Functions

In [ ]:
k = torch.randint(size=(2,3), low=0, high=10)
print(k)

# log
print(torch.log(k))

# exp
print(torch.exp(k))

# sqrt
print(torch.sqrt(k))

# sigmoid
print(torch.sigmoid(k))

# softmax
# print(torch.softmax(k, dim=0))

tensor([[9, 1, 7],
        [8, 6, 0]])
tensor([[2.1972, 0.0000, 1.9459],
        [2.0794, 1.7918,   -inf]])
tensor([[8.1031e+03, 2.7183e+00, 1.0966e+03],
        [2.9810e+03, 4.0343e+02, 1.0000e+00]])
tensor([[3.0000, 1.0000, 2.6458],
        [2.8284, 2.4495, 0.0000]])
tensor([[0.9999, 0.7311, 0.9991],
        [0.9997, 0.9975, 0.5000]])


## Copying tensor

In [ ]:
a = torch.rand(2,3)
print(a)

b = a
print(a)
print(b)

tensor([[0.6947, 0.0897, 0.8712],
        [0.1330, 0.4137, 0.6044]])
tensor([[0.6947, 0.0897, 0.8712],
        [0.1330, 0.4137, 0.6044]])
tensor([[0.6947, 0.0897, 0.8712],
        [0.1330, 0.4137, 0.6044]])


## Tensor Operations on GPU

In [ ]:
torch.cuda.is_available()

True

In [ ]:
device = torch.device('cuda')

In [ ]:
# Creating a new tensor on GPU
torch.rand((2,3), device=device)

tensor([[0.6130, 0.0101, 0.3984],
        [0.0403, 0.1563, 0.4825]], device='cuda:0')

In [ ]:
# Moving an existing tensor to GPU
a = torch.rand(2,3)

In [ ]:
a

tensor([[0.7581, 0.9037, 0.9555],
        [0.1035, 0.6258, 0.2849]])

This tensor a is created on CPU

In [ ]:
a.to(device)

tensor([[0.7581, 0.9037, 0.9555],
        [0.1035, 0.6258, 0.2849]], device='cuda:0')

Now, this a is transferred to and stored in GPU